File: FlashCards - py/FlashCards/guide.py
import sys
import cards

class Guide:
    def __init__(self) :
        self.launch()

    def launch(self) :
        self.introduction()
        print("Press: \n0 - View Flash Cards \n1 - Add Flash Cards \n2 - Edit Flash Cards \n3 - Quit Application")
        user_response = sys.stdin.readline().strip()        
        
        cards.Cards.make_category_file(self)
        self.do_action(user_response)
        self.conclusion()

    def do_action(self, action) :   
        if action == '0' :
            # view flashcards
            self.choose_category()
        elif action == '1' :
            # create flashcards
            self.add_flash_cards()
        elif action == '2' :
            # edit flashcards
            print("Editing Flash Cards")
        elif action == '3' :
            exit()
        else :
            print("\nThis command does not exist!")

    def add_flash_cards(self) :
        print("Press: \n0 - Add new category:")
        categories = cards.Cards.saved_categories(self)
        if categories :
            print("or Add cards to existing category:")
            for i, category in enumerate(categories):
                print (str(i+1) + " - " + category)
        
        user_response = sys.stdin.readline().strip()
        self.do_category(user_response)

    def do_category(self, action) :
        path = ""
        categories = cards.Cards.saved_categories(self)
        if action == '0' :
            self.new_category()
        elif categories :
            for i, category in enumerate(categories):
                i = i + 1
                if action == str(i) :
                    path = category
            cards.Cards.save_category_name(self, path)
            print("Saved path: " + str(path))
            self.add_card(path) 
            ##
        else :
            print("Command not understood") 
            self.launch()

    def new_category(self) :
        print("What would you like to call this category of flashcards?")
        path = sys.stdin.readline().strip().lower()

        cards.Cards.make_category_file(self)
        cards.Cards.save_category_name(self, path)
        self.add_card(path)

    def add_card(self, path) :
        print("\nAdd a card:")
        
        print("Front of card:")
        front = sys.stdin.readline().strip().lower()

        print("Back of card:")
        back = sys.stdin.readline().strip().lower()

        card = ("\t").join([front, back])

        if cards.Cards.save_card(self, path, card) :
            print("Card Added")
        else :
            print("Save error: Card not added")

        print("Add another card to this category? type: (yes || no)")  
        res = sys.stdin.readline().strip().lower()  
        if res == 'yes' :
            self.add_card(path)
        elif res == 'no' :
            self.launch()
        else :
            print("Type either 'yes' or 'no' to proceed")
            print("Add another card to this category?")
        
    def choose_category(self) :
        car = cards.Cards.saved_categories(self)
        
        if car :
            print("Choose a category:")
            for i, category in enumerate(car):
                print (str(i) + " - " + category)
        else :
            print("You haven't added anything yet!")
            self.add_flash_cards()

        user_response = sys.stdin.readline().strip().lower()
        self.view_cards(user_response)


    def view_cards(self, user_response) :
        path = ""
        cars = cards.Cards.saved_categories(self)
        if cars :
            print("View Category: " + user_response)
            for i, category in enumerate(cars):
                if user_response == str(i) :
                    path = str(category)
            
                    x = 0
                    self.play_cards(path, x)

    # def list_cards(self, path) :
    #     print("Listing cards in categeory: " + path)
    #     car = cards.Cards.saved_card(self, path)
    #     print(car[0].front)

    def play_cards(self, path, i) :
        car = cards.Cards.saved_card(self, path)

        f = "CARD FRONT: " + car[i].front
        print(f)

        print("Press 'c' to check back of card, Press '1' to Exit")
        user_response = sys.stdin.readline().strip().lower()

        if user_response == 'c' :
            b = "CARD BACK: " + car[i].back
            print(b)

            self.play_cards(path, i)

        elif user_response == '1' :
            self.launch()
        else:
            print("Command not found")



    def introduction(self) :
        print("\n\n<<< Welcome to Flash Cards >>>\n")
        print("Add words and definitions to help yourself learn new things")

    def conclusion(self) :
        print("\n<<< Goodbye! Exiting Flash Cards! >>>\n\n")

File: FlashCards - ruby/FlashCards/guide.rb
class Guide
          
    def initialize
        launch
        # Cards.filepath = path
        # if Cards.file_usable?
        #   puts "Found flashcard file."
        # # or create a new file
        # elsif Cards.create_file
        #   puts "Created flashcard file."
        # # exit if create fails
        # else
        #   puts "Exiting.\n\n"
        #   exit!
        # end
    end

    def launch
        introduction
        # action loop
        puts "Press: \n0 - View Flash Cards \n1 - Add Flash Cards \n2 - Edit Flash Cards \n3 - Quit Application \n\n"
        print "> "
        user_response = gets.chomp
        # do that action
        result = do_action(user_response)

        conclusion
    end

    def do_action(action)
        case action
        when '0'
            choose_category
        when '1'
            add_flash_cards
        when '2'
            puts "Editing flash cards..."
        when '3'
            return :quit
        else
            puts "\nThis command does not exist!\n"
        end
    end

    def add_flash_cards 
        puts "Press: \n0 - Add new category"
        if Cards.saved_categories.any?
            puts "or Add cards to existing category:\n"  
            cards = Cards.saved_categories
            cards.to_enum.with_index(1).each do |category, i|
                puts "#{i} - #{category}"
            end
        end

        print "> "
        user_response = gets.chomp
        result = do_category(user_response)
    end

    def do_category(action)
        path = nil
        if action == '0'
            new_category
        elsif Cards.saved_categories.any?
            cards = Cards.saved_categories
            cards.to_enum.with_index(1).each do |category, i|
                if action == "#{i}"
                    path = "#{category}"                
                end
            end
            Cards.filepath = path.downcase.strip
            puts "\nCatergory Selected: " + path
            add_card
        else
            puts "Command not understood"
        end
    end    

    def choose_category
        if Cards.saved_categories.any?
            puts "Choose a category:\n"  
            cards = Cards.saved_categories
            cards.to_enum.with_index(1).each do |category, i|
                puts "#{i} - #{category}"
            end
        end    
        
        print "> "
        user_response = gets.chomp
        result = view_cards(user_response)
    end

    def view_cards(action)
        path = nil
        if Cards.saved_categories.any?
            puts "View Category:" + action
            cards = Cards.saved_categories
            cards.to_enum.with_index(1).each do |category, i|
                if action == "#{i}"
                    path = "#{category}"                
                end
            end
            Cards.filepath = path.downcase.strip
            puts "\nCatergory Selected: \s" + path
            
            # view cards here
            i = 0  
            play_cards(i)

        else
            puts "Command not understood"
        end
    end       

    def list_cards(path)
        puts "\nListing cards in category: " + path
        cards = Cards.saved_card
        cards.each do |card|
            puts card.front + " | " + card.back
        end
    end

    def play_cards(i)
        cards = Cards.saved_card

        f = "CARD FRONT: " + cards[i].front
        puts f   

        puts "Press 'c' to check back of card, Press '1' to Exit"
        user_response = gets.chomp    
        
        if user_response == 'c' 
            b = "CARD BACK: " + cards[i].back + "\n"
            puts b  
            
            if cards[i + 1] == nil
                i = 0
            end
            play_cards(i) 
        elsif user_response == '1'
            guide = Guide.new
        else
            puts("Command not found")
        end

    end

    def play_card(cards, i)
        f = "CARD FRONT: " + cards[i].front
        puts f       
    end

    def new_category
        puts "What would you like to call this category of flashcards?"
        print "> "
        user_response = gets.chomp
        path = user_response.downcase.strip

        Cards.filepath = path
        if Cards.file_usable?
          puts "It looks like this category already exists!"
        # or create a new file
        elsif Cards.create_file
          puts "Created flashcard category named " + path
          Cards.save_category_name(path)
          list_categories
        # exit if create fails
        else
          puts "Exiting.\n\n"
          exit!
        end
    end

    def list_categories
        puts "Listing Categories"
        cards = Cards.saved_categories
        cards.each do |card|
            puts card
        end
    end

    def add_card
        puts "\nAdd a card\n"
        cards = Cards.new
        
        print "Front of card: "
        cards.front = gets.chomp.strip
    
        print "Back of card: "
        cards.back = gets.chomp.strip
        
        if cards.save_card
          puts "\nCard Added\n\n"
        else
          puts "\nSave Error: Card not added\n\n"
        end   
        
        puts "\nAdd another card to this category type: ('yes' || 'no') \n"
        answ = gets.chomp.strip

        if answ == 'yes'
            add_card
        else
            launch
        end
    end    
    
    def introduction
        puts "\n\n<<< Welcome to Flash Cards >>>\n\n"
        puts "Add words and definitions to help you learn new things\n\n"
    end
    
    def conclusion
        puts "\n<<< Goodbye! Exiting Flash Cards >>>\n\n\n"
    end    
end
